src/Spago/Config.purs

```diff
@@ -174,130 +175,212 @@ type ReadWorkspaceOptions =
-        , Log.break
-        , indent $ toDoc errLines
-        , Log.break
-        , toDoc "The configuration file help can be found here https://github.com/purescript/spago#the-configuration-file"
-        ]
-    Right { yaml: { workspace: Nothing } } -> die
-      [ "Your spago.yaml doesn't contain a workspace section."
-      , "See the relevant documentation here: https://github.com/purescript/spago#the-workspace"
-      ]
-    Right config@{ yaml: { workspace: Just workspace, package }, doc } -> do
-      logDebug "Read the root config"
-      doMigrateConfig (rootPath </> "spago.yaml") config
-      pure { workspace, package, workspaceDoc: doc }
-
-  logDebug "Gathering all the spago configs in the tree..."
-  otherConfigPaths <- liftAff $ Array.delete rootConfigPath <$> Glob.gitignoringGlob
-    { root: rootPath
-    , includePatterns: [ "**/spago.yaml" ]
-    , ignorePatterns: [ "**/node_modules/**", "**/.spago/**" ]
-    }
+  lockfile <-
+    loadLockfile { pureBuild: options.pureBuild, workspaceConfig: workspace.config, loadedPackages: packagesByName, rootPath }
 
-  unless (Array.null otherConfigPaths) do
-    logDebug $ [ toDoc "Found packages at these paths:", Log.indent $ Log.lines (map (toDoc <<< Path.quote) otherConfigPaths) ]
+  { packageSet, compiler } <-
+    loadPackageSet { workspaceConfig: workspace.config, loadedPackages: packagesByName, rootPath, lockfile }
 
-  -- We read all of them in, and only read the package section, if any.
-  let
-    readWorkspaceConfig :: LocalPath -> Spago (Registry.RegistryEnv _) (Either Docc ReadWorkspaceConfigResult)
-    readWorkspaceConfig path = do
-      maybeConfig <- readConfig path
-      -- We try to figure out if this package has tests - look for test sources
-      hasTests <- FS.exists (Path.dirname path </> "test")
-      pure $ case maybeConfig of
-        Left eLines -> Left $ toDoc
-          [ toDoc $ "Could not read config at path " <> Path.quote path
-          , toDoc "Error was: "
-          , indent $ toDoc eLines
-          ]
-        Right config -> Right
-          { config
-          , hasTests
-          , configPath: path
-          , packagePath: Path.dirname path `Path.relativeTo` rootPath
+  pure
+    { rootPath
+    , workspace:
+        { selected
+        , packageSet
+        , compatibleCompiler: compiler
+        , backend: workspace.config.backend
+        , buildOptions:
+            { output: workspace.config.buildOpts >>= _.output <#> \o -> withForwardSlashes $ rootPath </> o
+            , censorLibWarnings: _.censorLibraryWarnings =<< workspace.config.buildOpts
+            , statVerbosity: _.statVerbosity =<< workspace.config.buildOpts
+            }
+        , doc: Just workspace.doc
+        , workspaceConfig: workspace.config
+        , rootPackage: workspace.rootPackage
+        }
+    }
+  where
+  readConfig' :: ∀ s. _ -> State.StateT s _ _
+  readConfig' = State.lift <<< readConfig
+
+  walkDirectoriesUpFrom dir = do
```

Comment:
I find the State monad always super confusing - let's have full type signatures on these internal functions so they are easier to figure out

Reply:
I added the signatures, but I am not at all sure they make it any clearer. If anything, I think they just create noise. Let me know if you like them.

Reply:
My issue with State is that if the program is large enough then you lose track of what's modifying the State and your functional program becomes a side-effectful one (Refs have the same problem of course). I've been in a codebase like that and it's no fun. Having the type signatures on these funs helps to distinguish what is stateful and what is not.

Reply:
Ah, well, yes, that is a fair concern, which I share. And this is why these functions are small and limited in scope.